File: Framework/PythonInterface/mantid/api/src/Exports/MultipleExperimentInfos.cpp
// Mantid Repository : https://github.com/mantidproject/mantid
//
// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
//   NScD Oak Ridge National Laboratory, European Spallation Source,
//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
// SPDX - License - Identifier: GPL - 3.0 +
#include "MantidAPI/MultipleExperimentInfos.h"
#include "MantidAPI/ExperimentInfo.h"
#include "MantidAPI/Workspace.h"
#include "MantidPythonInterface/core/ExtractSharedPtr.h"
#include <boost/python/class.hpp>

using Mantid::API::ExperimentInfo;
using Mantid::API::ExperimentInfo_sptr;
using Mantid::API::MultipleExperimentInfos;
using Mantid::API::Workspace;
using Mantid::PythonInterface::ExtractSharedPtr;
using namespace boost::python;

namespace {
void addExperimentInfo(MultipleExperimentInfos &self, const boost::python::object &item) {
  auto workspaceExtractor = ExtractSharedPtr<Workspace>(item);
  if (!workspaceExtractor.check())
    throw std::invalid_argument("Incorrect type. Expected a workspace.");

  if (auto exptInfo = std::dynamic_pointer_cast<ExperimentInfo>(workspaceExtractor()))
    self.addExperimentInfo(exptInfo);
  else
    throw std::invalid_argument("Incorrect type. Expected a workspace type derived from ExperimentInfo.");
}
} // namespace

void export_MultipleExperimentInfos() {
  class_<MultipleExperimentInfos, boost::noncopyable>("MultipleExperimentInfos", no_init)
      .def(
          "getExperimentInfo",
          (ExperimentInfo_sptr (MultipleExperimentInfos::*)(const uint16_t))&MultipleExperimentInfos::getExperimentInfo,
          (arg("self"), arg("expInfoIndex")), "Return the experiment info at the given index.")
      .def("addExperimentInfo", addExperimentInfo, (arg("self"), arg("ExperimentalInfo")),
           "Add a new :class:`~mantid.api.ExperimentInfo` to this "
           ":class:`~mantid.api.IMDWorkspace`")
      .def("getNumExperimentInfo", &MultipleExperimentInfos::getNumExperimentInfo, arg("self"),
           "Return the number of :class:`~mantid.api.ExperimentInfo` objects,")
      .def("copyExperimentInfos", &MultipleExperimentInfos::copyExperimentInfos,
           (arg("self"), arg("MultipleExperimentInfos")), "Copy the :class:`~mantid.api.ExperimentInfo` from another");
}

File: Framework/PythonInterface/test/python/mantid/api/MultipleExperimentInfos.py
# Mantid Repository : https://github.com/mantidproject/mantid
#
# Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
#   NScD Oak Ridge National Laboratory, European Spallation Source,
#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
# SPDX - License - Identifier: GPL - 3.0 +
import unittest
from testhelpers import WorkspaceCreationHelper
from mantid.api import AnalysisDataService, ExperimentInfo
from mantid.simpleapi import CreateSampleWorkspace, CreateMDWorkspace


class MultipleExperimentInfoTest(unittest.TestCase):

    def tearDown(self):
        AnalysisDataService.clear()

    def test_information_access(self):
        signal = 2.0
        ndims = 1
        expt_ws = WorkspaceCreationHelper.makeFakeMDHistoWorkspace(signal, ndims)
        expinfo = expt_ws.getExperimentInfo(0)
        self.assertTrue(isinstance(expinfo, ExperimentInfo))
        self.assertEqual(1, expt_ws.getNumExperimentInfo())

    def test_add_experiment_info(self):
        ws1 = CreateSampleWorkspace()
        md = CreateMDWorkspace(Dimensions=1, Extents='-1,1', Names='A', Units='U')
        md.addExperimentInfo(ws1)

if __name__ == '__main__': unittest.main()
